File: goreview/internal/providers/groq.go
package providers

import (
	"context"
	"fmt"
	"net/http"
	"time"

	"github.com/JNZader/goreview/goreview/internal/config"
)

// GroqProvider implements Provider using Groq API.
// Groq uses OpenAI-compatible API format.
type GroqProvider struct {
	apiKey  string
	baseURL string
	model   string
	client  *http.Client
	config  *config.ProviderConfig
}

// NewGroqProvider creates a new Groq provider.
func NewGroqProvider(cfg *config.Config) (*GroqProvider, error) {
	if cfg.Provider.APIKey == "" {
		return nil, fmt.Errorf("Groq API key required (get free at console.groq.com)")
	}

	baseURL := cfg.Provider.BaseURL
	if baseURL == "" {
		baseURL = "https://api.groq.com/openai/v1"
	}

	model := cfg.Provider.Model
	if model == "" {
		model = "llama-3.3-70b-versatile"
	}

	return &GroqProvider{
		apiKey:  cfg.Provider.APIKey,
		baseURL: baseURL,
		model:   model,
		config:  &cfg.Provider,
		client:  &http.Client{Timeout: cfg.Provider.Timeout},
	}, nil
}

func (p *GroqProvider) Name() string { return "groq" }

func (p *GroqProvider) Review(ctx context.Context, req *ReviewRequest) (*ReviewResponse, error) {
	if empty, err := ValidateReviewInput(req); err != nil {
		return nil, err
	} else if empty {
		return &ReviewResponse{}, nil
	}

	start := time.Now()
	groqReq := BuildChatRequest(p.model, ReviewSystemPrompt, buildReviewPrompt(req), p.config.Temperature, p.config.MaxTokens, true)

	var result ChatCompletionResponse
	if err := DoJSONPost(ctx, p.client, p.baseURL+ChatCompletionsPath, groqReq, p.apiKey, &result); err != nil {
		return nil, fmt.Errorf("groq request failed: %w", err)
	}

	if result.Error != nil {
		return nil, fmt.Errorf("groq error: %s", result.Error.Message)
	}

	return ParseReviewContent(result.GetContent(), result.Usage.TotalTokens, time.Since(start).Milliseconds()), nil
}

func (p *GroqProvider) GenerateCommitMessage(ctx context.Context, diff string) (string, error) {
	groqReq := map[string]interface{}{
		"model":    p.model,
		"messages": []map[string]string{{"role": "user", "content": fmt.Sprintf(CommitMessagePrompt, diff)}},
	}

	var result ChatCompletionResponse
	if err := DoJSONPost(ctx, p.client, p.baseURL+ChatCompletionsPath, groqReq, p.apiKey, &result); err != nil {
		return "", err
	}

	if content := result.GetContent(); content != "" {
		return content, nil
	}
	return "", fmt.Errorf("no response from Groq")
}

func (p *GroqProvider) GenerateDocumentation(ctx context.Context, diff, docContext string) (string, error) {
	groqReq := map[string]interface{}{
		"model":    p.model,
		"messages": []map[string]string{{"role": "user", "content": fmt.Sprintf(DocumentationPrompt, docContext, diff)}},
	}

	var result ChatCompletionResponse
	if err := DoJSONPost(ctx, p.client, p.baseURL+ChatCompletionsPath, groqReq, p.apiKey, &result); err != nil {
		return "", err
	}

	if content := result.GetContent(); content != "" {
		return content, nil
	}
	return "", fmt.Errorf("no response from Groq")
}

func (p *GroqProvider) HealthCheck(ctx context.Context) error {
	return DoHealthCheck(ctx, p.client, p.baseURL+"/models", p.apiKey, "groq")
}

func (p *GroqProvider) Close() error { return nil }

File: github-app/src/services/ai/groq.ts
/**
 * Groq AI Provider.
 * Uses OpenAI-compatible API format.
 */

import { config } from '../../config/index.js';
import { logger } from '../../utils/logger.js';
import {
  AIProvider,
  ReviewRequest,
  ReviewResponse,
  buildReviewPrompt,
  buildCommitPrompt,
  parseReviewResponse,
} from './types.js';

export class GroqProvider implements AIProvider {
  readonly name = 'groq';
  private readonly baseUrl: string;
  private readonly model: string;
  private readonly apiKey: string;

  constructor() {
    this.apiKey = config.ai.groqApiKey || '';
    this.baseUrl = 'https://api.groq.com/openai/v1';
    this.model = config.ai.model || 'llama-3.3-70b-versatile';
  }

  async review(request: ReviewRequest): Promise<ReviewResponse> {
    const startTime = Date.now();
    const prompt = buildReviewPrompt(request);

    const groqReq = {
      model: this.model,
      messages: [
        { role: 'system', content: 'You are an expert code reviewer. Return valid JSON only.' },
        { role: 'user', content: prompt },
      ],
      temperature: 0.3,
      max_tokens: 4096,
      response_format: { type: 'json_object' },
    };

    logger.debug({ model: this.model }, 'Calling Groq API');

    const response = await fetch(`${this.baseUrl}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify(groqReq),
      signal: AbortSignal.timeout(config.review.timeout),
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Groq error: ${response.status} ${error}`);
    }

    const result = (await response.json()) as {
      choices?: Array<{
        message?: { content?: string };
      }>;
      usage?: {
        total_tokens?: number;
      };
      error?: {
        message: string;
        type: string;
      };
    };

    if (result.error) {
      throw new Error(`Groq error: ${result.error.message}`);
    }

    let reviewResponse: ReviewResponse;
    if (result.choices && result.choices[0]?.message?.content) {
      reviewResponse = parseReviewResponse(result.choices[0].message.content);
    } else {
      reviewResponse = { issues: [], summary: 'No response from Groq', score: 70 };
    }

    reviewResponse.tokensUsed = result.usage?.total_tokens;
    reviewResponse.processingTimeMs = Date.now() - startTime;

    return reviewResponse;
  }

  async generateCommitMessage(diff: string): Promise<string> {
    const prompt = buildCommitPrompt(diff);

    const groqReq = {
      model: this.model,
      messages: [{ role: 'user', content: prompt }],
    };

    const response = await fetch(`${this.baseUrl}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify(groqReq),
      signal: AbortSignal.timeout(config.review.timeout),
    });

    if (!response.ok) {
      throw new Error(`Groq error: ${response.status}`);
    }

    const result = (await response.json()) as {
      choices?: Array<{
        message?: { content?: string };
      }>;
    };

    if (result.choices && result.choices[0]?.message?.content) {
      return result.choices[0].message.content.trim();
    }

    throw new Error('No response from Groq');
  }

  async chat(prompt: string): Promise<string> {
    const groqReq = {
      model: this.model,
      messages: [
        { role: 'system', content: 'You are a helpful code review assistant. Be concise and helpful.' },
        { role: 'user', content: prompt },
      ],
      temperature: 0.7,
      max_tokens: 2048,
    };

    const response = await fetch(`${this.baseUrl}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify(groqReq),
      signal: AbortSignal.timeout(config.review.timeout),
    });

    if (!response.ok) {
      throw new Error(`Groq error: ${response.status}`);
    }

    const result = (await response.json()) as {
      choices?: Array<{
        message?: { content?: string };
      }>;
    };

    if (result.choices && result.choices[0]?.message?.content) {
      return result.choices[0].message.content.trim();
    }

    throw new Error('No response from Groq');
  }

  async healthCheck(): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/models`, {
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
        },
        signal: AbortSignal.timeout(5000),
      });
      return response.ok;
    } catch {
      return false;
    }
  }
}
